[FuncSpec] Only compute Latency bonus when necessary

hazzlim opened the pull request:
Only compute the Latency component of a specialisation's Bonus when necessary, to avoid unnecessarily computing the Block Frequency Information for a Function.

hazzlim committed:
[FuncSpec] Only compute Latency bonus when necessary

Only compute the Latency component of a specialisation's Bonus when
necessary, to avoid unnecessarily computing the Block Frequency
Information for a Function.

hazzlim commented:
Change in compile-times for CTMark (measured as instruction-count) when SpecializeLiteralConstant is enabled:

O3 (Non-LTO):

```
Benchmark        |  Change  |
-----------------+----------+
ClamAV           |  -0.03%  |
7zip             |  -0.01%  |
kimwitu++        |  -0.01%  |
tramp3d-v4       |   0.00%  |
sqlite3          |  -0.03%  |
mafft            |  -0.01%  |
SPASS            |  -0.01%  |
lencod           |  -0.01%  |
consumer-typeset |  -0.14%  |
Bullet           |  -0.01%  |
-----------------+----------+
GEOMEAN          |  -0.03%  |
```

ReleaseLTO-g:

```
Benchmark        |  Change  |
-----------------+----------+
ClamAV           |  -0.03%  |
7zip             |  +0.01%  |
kimwitu++        |  -0.02%  |
tramp3d-v4       |  +0.01%  |
sqlite3          |  -0.06%  |
mafft            |  +0.03%  |
SPASS            |  -0.01%  |
lencod           |   0.00%  |
consumer-typeset |  -0.08%  |
Bullet           |  +0.01%  |
-----------------+----------+
GEOMEAN          |  -0.01%  |
```

This patch has the largest effect on the consumer-typeset benchmark, which exhibits compile-time regressions (+0.14% for O3, +0.12% for ReleaseLTO-g) when SpecializeLiteralConstant is enabled despite no specializations being created.
With this patch applied the compile-time changes for consumer-typeset when SpecializeLiteralConstant is enabled are -0.01% for O3 and +0.04% for ReleaseLTO-g.

labrinea commented on llvm/include/llvm/Transforms/IPO/FunctionSpecialization.h:
```diff
@@ -192,26 +193,29 @@ class InstCostVisitor : public InstVisitor<InstCostVisitor, Constant *> {
   }
 
-  Bonus getSpecializationBonus(Argument *A, Constant *C);
+  Cost getCodeSizeBonus(Argument *A, Constant *C);
```
s/getCodeSizeBonus/getCodeSizeSavingsForArg

labrinea commented on llvm/include/llvm/Transforms/IPO/FunctionSpecialization.h:
```diff
@@ -192,26 +193,29 @@ class InstCostVisitor : public InstVisitor<InstCostVisitor, Constant *> {
-  Bonus getSpecializationBonus(Argument *A, Constant *C);
+  Cost getCodeSizeBonus(Argument *A, Constant *C);
+
+  Cost getCodeSizeBonusFromPendingPHIs();
```
s/getCodeSizeBonusFromPendingPHIs/getCodeSizeSavingsFromPendingPHIs

labrinea commented on llvm/include/llvm/Transforms/IPO/FunctionSpecialization.h:
```diff
@@ -192,26 +193,29 @@ class InstCostVisitor : public InstVisitor<InstCostVisitor, Constant *> {
+  Cost getCodeSizeBonusFromPendingPHIs();
 
-  Bonus getBonusFromPendingPHIs();
+  Cost getLatencyBonus();
```
s/getLatencyBonus/getLatencySavingsForKnownConstants

labrinea commented on llvm/include/llvm/Transforms/IPO/FunctionSpecialization.h:
```diff
@@ -192,26 +193,29 @@ class InstCostVisitor : public InstVisitor<InstCostVisitor, Constant *> {
 
-  Bonus getUserBonus(Instruction *User, Value *Use = nullptr,
-                     Constant *C = nullptr);
+  Cost getUserCodeSizeBonus(Instruction *User, Value *Use = nullptr,
```
s/getUserCodeSizeBonus/getCodeSizeSavingsForUser

hazzlim committed:
Address review comments

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -875,24 +885,23 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
-                        << ", Inlining = " << Score << "}\n");
+                        << CodeSize << ", Inlining = " << Score << "}\n");
 
+      Bonus B = {CodeSize, 0};
```
Fair point - I was keeping it as it seemed like a reasonable way to abstract away the getValue() logic and to pass to the lambda, but I see it's probably cleaner without. Done.

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -154,37 +154,55 @@ static Constant *findConstantFor(Value *V, ConstMap &KnownConstants) {
+  return CodeSize;
+}
+
+Cost InstCostVisitor::getLatencyBonus() {
```
Good point - I have added an explanatory comment.

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -889,18 +910,20 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
-      CodeSize += Visitor.getCodeSizeBonusFromPendingPHIs();
+      CodeSize += Visitor.getCodeSizeSavingsFromPendingPHIs();
 
       LLVM_DEBUG(dbgs() << "FnSpecialization: Specialization bonus {CodeSize = "
```
Can we sink this inside IsProfitable, so that all the debug messages about profitability are in the same place? Perhaps we can split it in three messages: one for Inlining, one for CodeSize and one for Latency.

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -889,18 +910,20 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
+      unsigned LatencySavings = 0;
+      unsigned CodeSizeSavings = getCostValue(CodeSize);
+      FunctionGrowth[F] += FuncSize - CodeSizeSavings;
 
```
Can we also sink all three of the above statements inside the lambda, to the point of their use? Since this isn't a read-only lambda anymore I think it makes sense to get rid of the parameters and capture the scope by reference.

hazzlim replied:
I've sunk/moved things into the lambda and refactored it to just capture the scope by reference.

Note - I've realized that we accumulate the CodeSize increase into FunctionGrowth[F] for all specializations that we evaluate, rather than just the actual specializations we create. I've added a TODO for fixing this, as I think the changes would be non-trivial and don't make sense to do in this PR.

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -929,11 +952,12 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
 
       // Create a new specialisation entry.
-      Score += std::max(B.CodeSize, B.Latency);
+      Score += std::max(CodeSizeSavings, LatencySavings);
```
I think we can move this inside the lamda too, just before the `return true` statement at the end.

hazzlim committed:
Address review comments 2

- Fix outdated function names in comments.
- Sink debug messages about profitability into the IsProfitable lambda,
  and also add percentage of Function size while we're here.
- Move other profitability related statements into IsProfitable.
- Add a TODO to improve the accumulation of codesize increases
  (FunctionGrowth)

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -819,6 +831,15 @@ static Function *cloneCandidateFunction(Function *F, unsigned NSpecs) {
+static unsigned getCostValue(const Cost &C) {
+  int64_t Value = *C.getValue();
+
+  assert(Value >= 0 && "CodeSize and Latency cannot be negative");
```
Yeah good point - I'm not sure what makes the most sense either. I've added an explanatory comment documenting the non-negative assumptions and how this is always true for the Cost kinds we use in FuncSpec (TCK_CodeSize and TCK_Latency).

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -154,37 +154,67 @@ static Constant *findConstantFor(Value *V, ConstMap &KnownConstants) {
+
+    uint64_t Weight = BFI.getBlockFreq(I->getParent()).getFrequency() /
+                      BFI.getEntryFreq().getFrequency();
+    Latency +=
```
Sorry I missed it before. It seems worthwhile to add a debug message here showing how much is the latency saving and which instruction it corresponds to.

labrinea commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -875,48 +909,67 @@ bool FunctionSpecializer::findSpecializations(Function *F, unsigned FuncSize,
+        unsigned CodeSizeSavings = getCostValue(CodeSize);
+        // TODO: We should only accumulate codesize increase of specializations
+        // that are actually created.
+        FunctionGrowth[F] += FuncSize - CodeSizeSavings;
```
It is true indeed. We only keep `std::min(NumCandidates * MaxClones, unsigned(AllSpecs.size()))` per invocation of `FunctionSpecializer::run()`. It's not trivial to know upfront whether a specialization will be selected even if this lambda says it's profitable. However, if we move the FunctionGrowth accumulation just before the lambda returns true, does it not improve the likelihood of being created? (at least we know that MinCodeSizeSavings and MinLatencySavings requirements are satisfied at that point)

labrinea replied:
Hmm, but we sometimes early exit due to inlining bonus. Bummer

labrinea replied:
Perhaps we can move it here https://github.com/llvm/llvm-project/blob/56f75b5c1510cea7648cad0cb32e4e0810edd0d9/llvm/lib/Transforms/IPO/FunctionSpecialization.cpp#L735. At that point we know F, we know FuncSize and we know CodeSizeSavings (it is S.Score as long as we remove `Score += std::max(CodeSizeSavings, LatencySavings)`). Can you try that?

hazzlim replied:
> Hmm, but we sometimes early exit due to inlining bonus. Bummer

Yeah, it's somewhat annoying - I've put it where it is for this reason.

>Perhaps we can move it here ...

I think that this would be the ideal place to do the accumulation to `FunctionGrowth[F]`.

Could you clarify what you mean by the following:
> (it is S.Score as long as we remove Score += std::max(CodeSizeSavings, LatencySavings)). 

The Score is initially the Inlining Bonus, and then we add `std::max(CodeSizeSavings, LatencySavings))` - do you mean to just simply set `Score = CodeSizeSavings` in the case where we reach this point i.e. we are specializing due to CodeSizeSavings/LatencySavings rather than the Inlining Bonus?

hazzlim replied:
Yeah actually - that still wouldn't work because we wouldn't be able to distinguish if the Score represented CodeSizeSavings or inlining bonus. I think we'd need to add something like `CodeSizeCost` to the `Spec` struct in order to pass this information through to where we actually do the specializations?

My feeling is that it would make the most sense to leave this as it is for this PR and make these changes in a follow up PR, if you agree?

labrinea replied:
Fair enough, it alredy an improvement and the problem you spotted is not introduced by this patch. Good we have a TODO.

hazzlim committed:
Address review 3

- Add debug statement to getLatencySavingsForKnownConstants

hazzlim commented:
@labrinea Thank you for reviewing these changes
